sematic/plugins/abstract_external_resource.py

```diff
@@ -457,3 +457,8 @@ def _with_status(self: T, state: ResourceState, message: str) -> T:
                 message=message,
             ),
         )
+
+    @classmethod
+    def cloud_requires_standalone(cls) -> bool:
```

Comment:
This bleeds Runner implementation and architecture details into the external resources. Also, this has to be overridden to be turned off in a case-by-case basis, which will realistically not happen for new external resources, as there is no mechanism to force developers to do this when it is the case.

Instead, how about replacing the `assert image_uri is not None` line with raising the exception you introduced?

Reply:
> - This bleeds Runner implementation and architecture details into the external resources.

No more than is already the case. External resources already have the ability to have different logic depending on whether they are running locally or in the cloud. That is actually a part of their design intent.

> Also, this has to be overridden to be turned off in a case-by-case basis, which will realistically not happen for new external resources, as there is no mechanism to force developers to do this when it is the case.

*We* are likely to be the only developers that ever write custom external resources. In the off case that any other developers ever do write their own, the default behavior is reasonable: if the resource is being instantiated in the cloud, it requires that this not happen inside the runner.

> Instead, how about replacing the assert image_uri is not None line with raising the exception you introduced?

For one thing, because as mentioned in the PR description, that error message does not get surfaced directly to the user. It is caught by a higher layer and wrapped in a more generic "Failed to create manifest" error. The failure from the `assert` does get logged as part of the full error trace for the broader exception, but that full trace only is logged on the server during remote resource activation, which is never surfaced directly to the user. The user only gets the "summary" error message from the outer layer passed to them because the server returns the str of the basic exception message via the API. With this approach, the user gets the full relevant stack trace and message surfaced in the run's logs in the UI as well as in the exception view.

Reply:
Please add a code comment to the location where you raise the error that it is raised there in order for the message and stack trace to be surfaced to the user.